metricbeat/module/jolokia/jmx/config.go

```diff
@@ -103,41 +104,258 @@ func (m AttributeMapping) Get(mbean, attr string) (Attribute, bool) {
 	return a, found
 }
 
-// Parse strings with properties with the format key=value, being:
-// - key a nonempty string of characters which may not contain any of the characters,
-//   comma (,), equals (=), colon, asterisk, or question mark.
-// - value a string that can be quoted or unquoted, if unquoted it cannot be empty and
-//   cannot contain any of the characters comma, equals, colon, or quote.
-var propertyRegexp = regexp.MustCompile("[^,=:*?]+=([^,=:\"]+|\".*\")")
+// MBeanName is an internal struct used to store
+// the information by the parsed `mbean` (bean name) configuration
+// field in `jmx.mappings`.
+type MBeanName struct {
+	Domain     string
+	Properties map[string]string
+}
 
-func canonicalizeMBeanName(name string) (string, error) {
-	// From https://docs.oracle.com/javase/8/docs/api/javax/management/ObjectName.html#getCanonicalName--
-	//
-	//   Returns the canonical form of the name; that is, a string representation where the
-	//   properties are sorted in lexical order.
-	//   The canonical form of the name is a String consisting of the domain part,
-	//   a colon (:), the canonical key property list, and a pattern indication.
-	//
-	parts := strings.SplitN(name, ":", 2)
+var mbeanRegexp = regexp.MustCompile("([^,=:*?]+)=([^,=:\"]+|\".*\")")
+
+// Canonicalize Returns the canonical form of the name; that is, a string representation where the
+// properties are sorted in lexical order.
+// The canonical form of the name is a String consisting of the domain part,
+// a colon (:), the canonical key property list, and a pattern indication.
+//
+// For more information refer to Java 8 [getCanonicalName()](https://docs.oracle.com/javase/8/docs/api/javax/management/ObjectName.html#getCanonicalName--)
+// method.
+//
+// Set "escape" parameter to true if you want to use the canonicalized name for a Jolokia HTTP GET request, false otherwise.
+func (m *MBeanName) Canonicalize(escape bool) string {
+
+	var propertySlice []string
+
+	r2 := strings.NewReplacer("\"", "!\"", ".", "!.", "!", "!!", "/", "!/")
```

Comment:
As with regexps, you can define this out of the function, and give it a more meaningful name :slightly_smiling_face: